FWCore/Modules/src/GenericConsumer.cc

```diff
@@ -1,23 +1,108 @@
+ *  by that module, including automatic copy-to-host products in the case of Alpaka-based modules.
+ *  If a branch name is used, all four fields must be present, separated by underscores; this module will depend only
+ *  on the matching product(s).
+ *  
+ *  Glob expressions ("?" and "*") are supported in module labels and within the individual fields of branch names,
+ *  similar to an OutputModule's "keep" statements.
+ *  Use "*" to depend on all products of a given category.
  */
 
 #include <algorithm>
 #include <string>
+#include <regex>
 #include <vector>
 
+#include <boost/algorithm/string/replace.hpp>
+
+#include "DataFormats/Provenance/interface/BranchDescription.h"
 #include "FWCore/Framework/interface/global/EDAnalyzer.h"
 #include "FWCore/Framework/interface/MakerMacros.h"
 #include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
 #include "FWCore/ParameterSet/interface/ParameterDescriptionNode.h"
 #include "FWCore/ParameterSet/interface/ParameterSet.h"
 #include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
 
+namespace {
+  struct ProductBranch {
+  public:
+    ProductBranch(std::string const& label) {
+      static const char kSeparator = '_';
+      static const char kWildcard = '*';
+      static const std::regex kAny{".*"};
+
+      // wildcard
+      if (label == kWildcard) {
+        type_ = kAny;
+        moduleLabel_ = kAny;
+        productInstanceName_ = kAny;
+        processName_ = kAny;
+        return;
+      }
+
+      int fields = std::count(label.begin(), label.end(), kSeparator) + 1;
+      if (fields == 1) {
+        // convert the module label into a regular expression
+        type_ = kAny;
+        moduleLabel_ = glob_to_regex(label);
+        productInstanceName_ = kAny;
+        processName_ = kAny;
+      } else if (fields == 4) {
+        // split the branch name into <product type>_<module label>_<instance name>_<process name>
+        // and convert the glob expressions into regular expressions
+        size_t first = 0, last = 0;
+        last = label.find(kSeparator, first);
+        type_ = glob_to_regex(label.substr(first, last - first));
+        first = last + 1;
+        last = label.find(kSeparator, first);
+        moduleLabel_ = glob_to_regex(label.substr(first, last - first));
+        first = last + 1;
+        last = label.find(kSeparator, first);
+        productInstanceName_ = glob_to_regex(label.substr(first, last - first));
+        first = last + 1;
+        last = label.find(kSeparator, first);
+        processName_ = glob_to_regex(label.substr(first, last - first));
+      } else {
+        // invalid input
+        throw edm::Exception(edm::errors::Configuration) << "Invalid module label or branoh name: \"" << label << "\"";
+      }
+    }
+
+    bool match(edm::BranchDescription const& branch) const {
+      return (std::regex_match(branch.friendlyClassName(), type_) and
```

Comment:
@makortel is `friendlyClassName()` the correct method for the product type as it's use by an OutputModule ?

Reply:
Yes, it is the correct one.

@Dr15Jones Do we have any facilities in the ProductSelectors (that the OutputModules use) that could be re-used here? By quick look that does not seem to be the case (so perhaps something for future refactoring).